Radius of gyration method

maurobedoya opened the pull request:
This pull request adds a simple method to calculate the radius of gyration of a molecule using the coords. The new method is called `rdgyr()`, and it takes the `structure.coords` as input.

maurobedoya committed:
Add radius of gyration method

maurobedoya committed:
Add file to spec

maurobedoya committed:
Delete test_rdgyr.cr

maurobedoya committed:
Fix shard.yml

franciscoadasme reviewed (changes requested):
Thank you for the PR! Everything seems OK except a few small issues (see specific comments). Also, please use the default Crystal formatter to avoid code styling inconsistencies (e.g., using empty parenthesis).

franciscoadasme commented on spec/spatial/coordinates_proxy_spec.cr:
```diff
@@ -192,6 +192,13 @@ describe Chem::Spatial::CoordinatesProxy do
+  describe ".rdgyr" do 
+    it "computes the rdgyr in-place" do 
+      s = Chem::Structure.read spec_file("FAD_can_prep.pdb")
+      s.coords.rdgyr().should be_close 6.339020, 1e-6
```
Avoid empty parenthesis. Use default Crystal formatter to avoid this issues.

```suggestion
      s.coords.rdgyr.should be_close 6.339020, 1e-6
```

franciscoadasme commented on src/chem/spatial/coordinates_proxy.cr:
```diff
@@ -251,6 +251,24 @@ module Chem::Spatial
+    # ri is the distance of each atom from the center of mass of the molecule
+    def rdgyr() : Float64
+      center = self.com
+      pos = to_a           # FIXME: avoid copying coordinates
```
No need to convert to an array as the coordinates are used directly

maurobedoya replied:
Thanks for the clarification!

franciscoadasme commented on src/chem/spatial/coordinates_proxy.cr:
```diff
@@ -251,6 +251,24 @@ module Chem::Spatial
+    def rdgyr() : Float64
+      center = self.com
+      pos = to_a           # FIXME: avoid copying coordinates
+      square_sum = pos.sum do |i|
```
Use `Enumerable#mean` method instead and remove `square_sum/pos.size` below.

franciscoadasme commented on src/chem/spatial/coordinates_proxy.cr:
```diff
@@ -251,6 +251,24 @@ module Chem::Spatial
+    def rdgyr() : Float64
+      center = self.com
+      pos = to_a           # FIXME: avoid copying coordinates
+      square_sum = pos.sum do |i|
+        d = center.distance(i)
+        d.abs2
+      end
+      Math.sqrt(square_sum/pos.size)
```
This can be written more succinctly  by using `Enumerable#mean` and `Vec3#distance2(Vec3)` methods:

```suggestion
      Math.sqrt pos.mean(&.distance2(center))
```

franciscoadasme commented on src/chem/spatial/coordinates_proxy.cr:
```diff
@@ -251,6 +251,24 @@ module Chem::Spatial
       end
     end
 
+    # Returns the radius of gyration of a structure (RDGYR) in Å.
+    #
+    # The RDGYR is the root-mean-square distance of the atoms of the 
+    # molecule from the axis of rotation. The RDGYR is a measure of the 
+    # distribution of its atoms around its center of mass.
+    # RDGYR =  √(1/N * Σ(ri²))
+    # N = number of atoms in the molecule
+    # ri is the distance of each atom from the center of mass of the molecule
```
Please avoid using abbreviations in the docs unless they're standard. I suggest flipping the two sentences such that the formula follows the definition, and writing using full paragraphs and sentences.

```suggestion
    # Returns the radius of gyration in Å.
    #
    # The radius of gyration is a measure of the distribution of the
    # atoms around the center of mass of a molecule. The radius of
    # gyration is defined as the root-mean-square distance of the
    # atoms from the axis of rotation:
    #
    # RDGYR =  √ 1 / N * Σ(r - c)²
    #
    # where *N* is the number of atoms in the molecule, $r$ is the
    # coordinates of each atom, and $c$ is the center of mass of the 
    # molecule.
```

maurobedoya committed:
Refactor spec/spatial/coordinates_proxy_spec.cr

Co-authored-by: Francisco Adasme <franciscoadasme@users.noreply.github.com>

maurobedoya committed:
Refactor src/chem/spatial/coordinates_proxy.cr

Co-authored-by: Francisco Adasme <franciscoadasme@users.noreply.github.com>

maurobedoya committed:
Update src/chem/spatial/coordinates_proxy.cr

Co-authored-by: Francisco Adasme <franciscoadasme@users.noreply.github.com>

maurobedoya reviewed (commented):
Thanks for the review! The comments and modifications were very clarifying.
Happy to contribute!

franciscoadasme commented on src/chem/spatial/coordinates_proxy.cr:
```diff
@@ -251,6 +251,24 @@ module Chem::Spatial
+    #
+    # RDGYR =  √ 1 / N * Σ(r - c)²
+    #
+    # where *N* is the number of atoms in the molecule, $r$ is the
+    # coordinates of each atom, and $c$ is the center of mass of the 
```
Oops, it should be `*r*`, not `$r$`.

```suggestion
    # where *N* is the number of atoms in the molecule, *r* is the
    # coordinates of each atom, and *c* is the center of mass of the 
```

maurobedoya committed:
Fix symbol in coordinates_proxy.cr

Co-authored-by: Francisco Adasme <franciscoadasme@users.noreply.github.com>

maurobedoya committed:
Refactor coordinates_proxy.cr

maurobedoya committed:
Fix method to take directly positions without array creation

This fix increase the performance of the test from 2.96 ms to 2.66 ms.
Use as: crystal spec spec/spatial/coordinates_proxy_spec.cr:195 --release
